flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGeneratorContext.scala

```diff
@@ -481,6 +511,79 @@ class CodeGeneratorContext(val tableConfig: TableConfig) {
     fieldTerm
   }
 
+  /**
+   * Adds a reusable session-level time to the beginning of the SAM of
+   * the generated [[Function]].
+   */
+  def addReusableSessionTime(): String = {
+    val fieldTerm = s"sessionTime"
+
+    val sessionTimestamp = addReusableSessionTimestamp()
+    reusableMemberStatements.add(
+      s"""
+      |private static final int $fieldTerm = (int)
+      | ($sessionTimestamp.getMillisecond() % ${DateTimeUtils.MILLIS_PER_DAY}) < 0 ?
```

Comment:
I think the optimization can work only If we agree there's no case that current millisecond will be small than zero, I tend to keep the check because if user set wrong JVM time flink should still work